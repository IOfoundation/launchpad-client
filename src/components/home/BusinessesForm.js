import React from 'react';
import {PropTypes} from 'prop-types';
import onClickOutside from 'react-onclickoutside';
import {isEmpty} from 'lodash';
import {Link} from 'react-router';

class BusinessesForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      searchText: '',
      showDropdown: false,
      showPreviewDropdown: false,
    };
  }

  handleClickOutside() {
    this.setState({
      searchText: '',
      showDropdown: false,
      showPreviewDropdown: false,
    });
    const {value} = this.state;
    this.props.getTextSearchResults(value);
  }

  handleKeyPress(event) {
    const value = event.target.value;
    this.setState({
      searchText: value,
      showDropdown: true,
      showPreviewDropdown: false,
    });
    this.props.getTextSearchResults(value);
    if (isEmpty(value)) {
      this.setState({showDropdown: false});
    }
  }
  inputOnClick() {
    this.setState({showPreviewDropdown: true});
  }
  defaultDropdownOptions() {
    return (
      <ul className="hero-dropdown-list">
        <li>
          <Link to="/businesses?category=Planning/Management">
            {'Business Planning/Management'}
          </Link>
        </li>
        <li>
          <Link to="/businesses?category=Capital">{'Capital'}</Link>
        </li>
        <li>
          <Link to="/businesses?category=Legal%20Services">
            {'Legal Services'}
          </Link>
        </li>
        <li>
          <Link to="/businesses?category=Marketing/Sales">
            {'Marketing/Sales'}
          </Link>
        </li>
        <li>
          <Link to="/businesses?category=Physical%20Space">
            {'Physical Space'}
          </Link>
        </li>
      </ul>
    );
  }
  renderDropdown() {
    return (
      <div>
        <ul className={
            this.state.showPreviewDropdown
              ? 'hero-dropdown-list-hide'
              : 'hero-dropdown-list'
          }
        >
          {this.props.items.map(item => (
            <li
              key={item.id}
              className={item.searchable_type === 'Organization' ?
                'text-thin' : 'text-regular'
            }>
              <a
                href={
                  item.searchable_type === 'Category'
                    ? `/businesses?category=${item.content}`
                    : `/businesses?id=${item.searchable_id}`
                }
              >
                {item.content}
              </a>
              {item.searchable_type === 'Organization' ? (
                <img src="../static-data/images/organization.png" />
              ) : null}
            </li>
          ))}
          {this.props.items.length === 0 && (
            <li className="not-match-message">
              {'This does not match any item on our platform'}
            </li>
          )}
        </ul>
        {this.state.showPreviewDropdown ? this.defaultDropdownOptions() : null}
      </div>
    );
  }

  render() {
    return (
      <form>
        <input
          type='text'
          value={this.state.searchText}
          onChange={event => this.handleKeyPress(event)}
          onClick={() => this.inputOnClick()}
          placeholder={
            this.state.showDropdown || this.state.showPreviewDropdown
            ? ''
            : 'Search for businesses or services'
          }
          className="hero_input businessesName text-thin"
        />
        <img
          className="text-search-icon"
          src="/static-data/images/search.png"
        />
        <div
          className={
            this.state.showDropdown || this.state.showPreviewDropdown
              ? 'hero_input-dropdown hero_input-show'
              : 'hero_input-dropdown hero_input-hide'
          }
        >
          {this.renderDropdown()}
        </div>
      </form>
    );
  }
}

BusinessesForm.propTypes = {
  getTextSearchResults: PropTypes.func.isRequired,
  items: PropTypes.arrayOf(PropTypes.object),
};

export default onClickOutside(BusinessesForm);
