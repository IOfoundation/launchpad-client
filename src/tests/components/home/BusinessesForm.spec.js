import React from 'react';
import {mount} from 'enzyme';
import toJson from 'enzyme-to-json';

import BusinessesForm from '../../../components/home/BusinessesForm';

const props = {
  getTextSearchResults: jest.fn(),
  items: [{id: 1, content: 'content', searchable_type: 'Category'}],
};

describe('<BusinessesForm />', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(<BusinessesForm {...props} />);
  });

  it('Search box renders results', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Should call set the correct placeholder', () => {
    expect(wrapper.find('input').prop('placeholder')).toEqual(
      'Search all businesses and services'
    );
  });

  xit('Should displasy the dropdown', () => {
    expect(wrapper.find('.hero_input-dropdown').prop('className')).toEqual(
      'hero_input-dropdown hero_input-hide'
    );

    wrapper.find('.text-search-icon').prop('onClick')();

    expect(wrapper.find('.hero_input-dropdown').prop('className')).toEqual(
      'hero_input-dropdown hero_input-show'
    );
  });

  it('Should get the text search results', () => {
    wrapper.find('input').simulate('change', {target: {value: 'content'}});

    expect(props.getTextSearchResults).toHaveBeenCalled();
  });
});
