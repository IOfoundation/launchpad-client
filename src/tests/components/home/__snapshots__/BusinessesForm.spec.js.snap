// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BusinessesForm /> Search box renders results 1`] = `
<OnClickOutside(BusinessesForm)
  eventTypes={
    Array [
      "mousedown",
      "touchstart",
    ]
  }
  excludeScrollbar={false}
  getTextSearchResults={[MockFunction]}
  items={
    Array [
      Object {
        "content": "content",
        "id": 1,
        "searchable_type": "Category",
      },
    ]
  }
  outsideClickIgnoreClass="ignore-react-onclickoutside"
  preventDefault={false}
  stopPropagation={false}
>
  <BusinessesForm
    disableOnClickOutside={[Function]}
    enableOnClickOutside={[Function]}
    eventTypes={
      Array [
        "mousedown",
        "touchstart",
      ]
    }
    getTextSearchResults={[MockFunction]}
    items={
      Array [
        Object {
          "content": "content",
          "id": 1,
          "searchable_type": "Category",
        },
      ]
    }
    outsideClickIgnoreClass="ignore-react-onclickoutside"
    preventDefault={false}
    stopPropagation={false}
  >
    <form
      className="businesses-form"
      onKeyDown={[Function]}
      onMouseMove={[Function]}
    >
      <input
        className="hero_input text-thin"
        onChange={[Function]}
        onClick={[Function]}
        placeholder="Search for businesses and services"
        type="text"
        value=""
      />
      <img
        className="text-search-icon"
        onClick={[Function]}
        src="/static-data/images/search.png"
      />
      <div
        className="hero_input-dropdown hero_input-hide"
      >
        <div>
          <ul
            className="hero-dropdown-list scroll-list"
          >
            <li
              className="text-regular"
              key="1"
              onMouseOver={[Function]}
              value={1}
            >
              <a
                href="/businesses?category=content"
              >
                content
              </a>
            </li>
          </ul>
        </div>
      </div>
    </form>
  </BusinessesForm>
</OnClickOutside(BusinessesForm)>
`;
